module/zfs/zpl_inode.c

```diff
@@ -226,6 +226,11 @@ zpl_tmpfile(struct inode *dir, struct dentry *dentry, zpl_umode_t mode)
 
 	crhold(cr);
 	vap = kmem_zalloc(sizeof (vattr_t), KM_SLEEP);
+	/*
+	 * The VFS does not apply the umask to the provided mode as of 5.2.
+	 * Therefore, it must be applied at the filesystem layer.
```

Comment:
Was this really a change in the 5.2 kernel, or does it predate it?  That shouldn't matter for the fix itself, but I'm curious.

Reply:
This has nothing to do with version 5.2, this bug is a lot older, and may even be older than `O_TMPFILE` itself.

Reply:
btw. the rest of the Linux kernel applies the umask only if there's no ACL, e.g.:
https://github.com/torvalds/linux/blob/v5.2/fs/posix_acl.c#L596-L597
https://github.com/torvalds/linux/blob/v5.2/fs/namei.c#L3176-L3177
This PR applies it unconditionally.

Reply:
> Was this really a change in the 5.2 kernel, or does it predate it? 

I meant to say it has been there and still not supported by VFS in 5.2.

Reply:
Thanks for pointing out that the kernel only applies this when there is no ACL.  @kusumi what do you think about the follow modification.

```diff --git a/module/zfs/zpl_inode.c b/module/zfs/zpl_inode.c
index 3f3b2e2..9854791 100644
--- a/module/zfs/zpl_inode.c
+++ b/module/zfs/zpl_inode.c
@@ -226,6 +226,15 @@ zpl_tmpfile(struct inode *dir, struct dentry *dentry, zpl_u
 
        crhold(cr);
        vap = kmem_zalloc(sizeof (vattr_t), KM_SLEEP);
+       /*
+        * The VFS does not apply the umask, therefore it is applied here
+        * when POSIX ACLs are not enabled.
+        */
+       if (!IS_POSIXACL(dir->d_inode))
+               mode &= ~current_umask();
+
        zpl_vap_init(vap, dir, mode, cr);
 
        cookie = spl_fstrans_mark();

```

Reply:
Then I guess we also want to change `ip->i_mode &= ~current_umask();` in `zpl_init_acl()` side too ? It sounds like ZoL was doing opposite of what VFS expects.

> zfs seems to be ignoring umask if files created with O_TMPFILE unless acltype=posixacl

Reply:
I'll check my diff and Linux VFS code again soon. If needed I'll make the change and rebase again.

Reply:
Applied `!IS_POSIXACL()` conditional + rebased.